src/app/qgsvectorlayersavestyledialog.cpp

```diff
@@ -0,0 +1,196 @@
+/***************************************************************************
+  qgsvectorlayersavestyledialog.h
+  --------------------------------------
+  Date                 : September 2018
+  Copyright            : (C) 2018 by Denis Rouzaud
+  Email                : denis@opengis.ch
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include <QListWidgetItem>
+#include <QMessageBox>
+
+#include "qgsvectorlayersavestyledialog.h"
+#include "qgsvectorlayer.h"
+#include "qgssettings.h"
+#include "qgshelp.h"
+
+QgsVectorLayerSaveStyleDialog::QgsVectorLayerSaveStyleDialog( QgsVectorLayer *layer, QWidget *parent )
+  : QDialog( parent )
+  , mLayer( layer )
+{
+  setupUi( this );
+  QgsSettings settings;
+
+  QString providerName = mLayer->providerType();
+  if ( providerName == QLatin1String( "ogr" ) )
+  {
+    providerName = mLayer->dataProvider()->storageType();
+    if ( providerName == QLatin1String( "GPKG" ) )
+      providerName = QStringLiteral( "GeoPackage" );
+  }
+
+  QString myLastUsedDir = settings.value( QStringLiteral( "style/lastStyleDir" ), QDir::homePath() ).toString();
+
+  // save style type combobox
+  connect( mStyleTypeComboBox, qgis::overload<int>::of( &QComboBox::currentIndexChanged ), this, [ = ]( int idx )
+  {
+    Q_UNUSED( idx );
+    QgsVectorLayerProperties::StyleType type = currentStyleType();
+    mSaveToFileWidget->setVisible( type != QgsVectorLayerProperties::DB );
+    mSaveToDbWidget->setVisible( type == QgsVectorLayerProperties::DB );
+    mStyleCategoriesListWidget->setEnabled( type == QgsVectorLayerProperties::QML );
+    mFileWidget->setFilter( type == QgsVectorLayerProperties::QML ? tr( "QGIS Layer Style File (*.qml)" ) : tr( "SLD File (*.sld)" ) );
+  } );
+  mStyleTypeComboBox->addItem( tr( "as QGIS QML style file" ), QgsVectorLayerProperties::QML );
```

Comment:
HIG guidelines is that combobox entries are title case: "As QGIS QML Style File"